### PR TITLE
Replace AssertionError with ReconcileError

### PR DESCRIPTION
**Pull request by HomayoonAlimohammadi**
### Overview

Replace `AssertionError` with a newly defined `ReconcileError`.

### CODE REVIEW

File: charms/worker/k8s/src/errors.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Custom exceptions for the worker charm."""
+
+class ReconcileError(Exception):
+    """Exception raised for errors that occur during the reconciliation process."""
+    ...
```
Comment (addyess) [inline charms/worker/k8s/src/errors.py:8]: hold up... this exception is already defined...

you'll find it here 
```from charms.contextual_status import ReconcilerError```


I think the ask here is to not use this style in the code:

 ```python
assert <truthy statement>, "Message about the truthy statement"
```

but rather:

```python
if not <truthy_statement>:
    raise ReconcilerError("Message about the truthy statement")
```

**Review by addyess — CHANGES_REQUESTED** [review]
Hey @HomayoonAlimohammadi thanks for taking a stab at this.  Let me help clarify the ask on this task

File: charms/worker/k8s/src/errors.py (hunk 1 above)
Comment (HomayoonAlimohammadi) [reply charms/worker/k8s/src/errors.py]: That's makes a lot more sense...
Done. Thanks!

File: charms/worker/k8s/src/charm.py [hunk 2]
```diff
@@ -184,7 +184,7 @@ def _k8s_info(self, event: ops.EventBase):
     @status.on_error(
         ops.WaitingStatus("Installing COS requirements"),
         subprocess.CalledProcessError,
-        AssertionError,
+        ReconcilerError,
```
Comment (mateoflorido) [inline charms/worker/k8s/src/charm.py:187]: Hmmm, this method doesn't raise a ReconcilerError, so it should be fine to remove it.

**Review by mateoflorido — APPROVED** [review]
LGTM! Just a nitpick

File: charms/worker/k8s/src/charm.py (hunk 2 above)
Comment (HomayoonAlimohammadi) [reply charms/worker/k8s/src/charm.py]: Thanks! Done!

File: charms/worker/k8s/src/token_distributor.py [hunk 3]
```diff
@@ -246,14 +257,17 @@ def recover_token(self, relation: ops.Relation) -> Generator[str, None, None]:
...
-        assert secret_id, f"{relation.name}:{secret_key} is not valid"  # nosec
+        if not secret_id:
+            raise ReconcilerError(f"{relation.name}:{secret_key} is not valid")
         secret = self.charm.model.get_secret(id=secret_id)
 
         # Get the content from the secret
         content = secret.get_content(refresh=True)
-        assert content["token"], f"{relation.name}:token not valid"  # nosec
+        if content.get("token") is None:
+            raise ReconcilerError(f"{relation.name}:token not valid")
```
Comment (addyess) [inline charms/worker/k8s/src/token_distributor.py:270]: i think this is more accurate

```suggestion
        if not content.get("token"):
            raise ReconcilerError(f"{relation.name}:token not valid")
```

**Review by addyess — APPROVED** [review]
Please fix tiny nit... then merge with passing tests

File: charms/worker/k8s/src/token_distributor.py (hunk 3 above)
Comment (HomayoonAlimohammadi) [reply charms/worker/k8s/src/token_distributor.py:270]: Thanks! Fixed this. Just as a question: Are we changing this to also include content.get("token") == ""?

**Pull request merged**